src/node/txreconciliation.cpp

```diff
@@ -0,0 +1,192 @@
+     *       They should be made private once they are used in one of the following commits.
+     */
+    const uint64_t m_k0, m_k1;
+
+    /**
+     * Reconciliation protocol assumes using one role consistently: either a reconciliation
+     * initiator (requesting sketches), or responder (sending sketches). This defines our role.
+     * */
+    const bool m_we_initiate;
+
+    ReconciliationState(uint64_t k0, uint64_t k1, bool we_initiate) :
+        m_k0(k0), m_k1(k1), m_we_initiate(we_initiate) {}
+};
+
+} // namespace
+
+/** Actual implementation for TxReconciliationTracker's data structure. */
+class TxReconciliationTracker::Impl {
+
+    mutable Mutex m_mutex;
+
+    /**
+     * Per-peer salt is used to compute transaction short IDs, which will be later used to
+     * construct reconciliation sketches.
+     * Salt is generated randomly per-peer to prevent:
+     * - linking of network nodes belonging to the same physical node
+     * - halting of relay of particular transactions due to short ID collisions (DoS)
+     */
+    std::unordered_map<NodeId, uint64_t> m_local_salts GUARDED_BY(m_mutex);
+
+    /**
+     * Keeps track of ongoing reconciliations with a given peer.
+     */
+    std::unordered_map<NodeId, ReconciliationState> m_states GUARDED_BY(m_mutex);
+
+    public:
+
+    std::tuple<bool, bool, uint32_t, uint64_t> PreRegisterPeer(NodeId peer_id, bool peer_inbound)
+    {
+        bool we_initiate_recon, we_respond_recon;
+        // Currently reconciliation roles are defined by the connection direction: only the inbound
+        // peer initiate reconciliations and the outbound peer is supposed to only respond.
+        // If the way we assign reconciliation roles is updated, RECON_VERSION should be bumped.
+        if (peer_inbound) {
+            we_initiate_recon = false;
+            we_respond_recon = true;
+        } else {
+            we_initiate_recon = true;
+            we_respond_recon = false;
+        }
+
+        uint64_t m_local_recon_salt(GetRand(UINT64_MAX));
+        bool added = WITH_LOCK(m_mutex, return m_local_salts.emplace(peer_id, m_local_recon_salt).second);
+        // We do this exactly once per peer (which are unique by NodeId, see GetNewNodeId) so it's
+        // safe to assume we don't have this record yet.
+        assert(added);
+
+        LogPrint(BCLog::NET, "Pre-register peer=%d for reconciling.\n", peer_id);
+        return std::make_tuple(we_initiate_recon, we_respond_recon, RECON_VERSION, m_local_recon_salt);
+    }
+
+    bool RegisterPeer(NodeId peer_id, bool peer_inbound,
+        bool they_may_initiate, bool they_may_respond, uint32_t recon_version, uint64_t remote_salt)
+    {
+        // We do not support reconciliation salt/version updates. We treat an attempt to update
+        // after a successful registration as a protocol violation.
+        LOCK(m_mutex);
+        if (m_states.find(peer_id) != m_states.end()) return false;
```

Comment:
Since 103e52dcd316a5b3e939f3880da566430820e567, I think this can be replace with IsPeerRegistered?

Reply:
I thought it's better to use internal access internally. Loose coupling & high cohesion i guess.

Reply:
I don't see how this helps with reducing coupling/increasing cohesion, when IsPeerRegistered is a public member of the same class? This just looks like a regular case of DRY imo. Plus, readability improves.

Reply:
Probably you are right